### PR TITLE
feat: [CDS-58338]: Adding Cluster name for K8S AWS,Azure,GCP connector type for instance sync V2 and fixing the cleanUpInstanceSyncPerpetualTaskInfo

### PR DESCRIPTION
**Pull request by namantalaycha10**
## Describe your changes
Adding Cluster name for K8S AWS,Azure,GCP connector type for instance sync V2 and fixing the cleanUpInstanceSyncPerpetualTaskInfo

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48454)

### CODE REVIEW

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/beans/InfrastructureOutcome.java [hunk 1]
```diff
@@ -30,4 +31,8 @@ public interface InfrastructureOutcome extends Outcome, PassThroughData, Executi
   String getInfraIdentifier();
   String getInfraName();
   Connector getConnector();
+
+  default String getCluster() throws UnsupportedTypeException {
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/beans/InfrastructureOutcome.java:35]: Any specific reason why we need this method? Not all infrastructure outcomes has cluster or concept of it in entity. And what about resource group and subscription id? Will we have to do the same? Why to not do casting to the actual type?

File: 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java [hunk 2]
```diff
@@ -143,10 +150,23 @@ public DeploymentInfoDTO getDeploymentInfo(
     K8sServerInstanceInfo k8sServerInstanceInfo = (K8sServerInstanceInfo) serverInstanceInfoList.get(0);
     LinkedHashSet<String> namespaces = getNamespaces(serverInstanceInfoList);
 
+    K8sCloudConfigMetadata k8sCloudConfigMetadata = null;
+    if ((infrastructureOutcome instanceof K8sGcpInfrastructureOutcome)
+        || (infrastructureOutcome instanceof K8sAwsInfrastructureOutcome)
+        || (infrastructureOutcome instanceof K8sAzureInfrastructureOutcome)) {
+      try {
+        k8sCloudConfigMetadata =
+            K8sAWSCloudConfigMetadata.builder().clusterName(infrastructureOutcome.getCluster()).build();
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java:159]: Why we're creating always AWS cloud config?

File: 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java [hunk 3]
```diff
@@ -91,12 +95,15 @@ public DeploymentReleaseDetails getDeploymentReleaseDetails(
...
-        k8sDeploymentReleaseDetailsList.add(K8sDeploymentReleaseDetails.builder()
-                                                .releaseName(k8sDeploymentInfoDTO.getReleaseName())
-                                                .namespaces(k8sDeploymentInfoDTO.getNamespaces())
-                                                .build());
+        k8sDeploymentReleaseDetailsList.add(
+            K8sDeploymentReleaseDetails.builder()
+                .releaseName(k8sDeploymentInfoDTO.getReleaseName())
+                .namespaces(k8sDeploymentInfoDTO.getNamespaces())
+                .k8sCloudClusterConfig(KubernetesCloudClusterConfig.builder()
+                                           .clusterName(k8sDeploymentInfoDTO.getCloudConfigMetadata().clusterName())
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java:104]: What about resource group and subscription id?

File: 126-instance/src/main/java/io/harness/helper/K8sAzureCloudConfigMetadata.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import io.harness.annotations.dev.OwnedBy;
+
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+@OwnedBy(HarnessTeam.CDP)
+public class K8sAzureCloudConfigMetadata implements K8sCloudConfigMetadata {
+  String clusterName;
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/helper/K8sAzureCloudConfigMetadata.java:20]: What about resource group and subscription id?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Comment by namantalaycha10** [comment]
trigger messageCheck

File: 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java [hunk 5]
```diff
@@ -143,13 +166,38 @@ public DeploymentInfoDTO getDeploymentInfo(
...
+      K8sAzureInfrastructureOutcome k8sAzureInfrastructureOutcome =
+          (K8sAzureInfrastructureOutcome) infrastructureOutcome;
+      return K8sAzureCloudConfigMetadata.builder()
+          .clusterName(k8sAzureInfrastructureOutcome.getCluster())
+          .resourceGroup(k8sAzureInfrastructureOutcome.getResourceGroup())
+          .subscription(k8sAzureInfrastructureOutcome.getSubscription())
+          .useClusterAdminCredentials(k8sAzureInfrastructureOutcome.getUseClusterAdminCredentials())
+          .build();
+    } else {
+      throw new InvalidArgumentsException(Pair.of("infrastructureOutcome",
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java:197]: We should never throw the exception and just return null

File: 126-instance/src/main/java/io/harness/helper/InstanceSyncHelper.java [hunk 6]
```diff
@@ -41,11 +43,18 @@ public class InstanceSyncHelper {
...
+      InstanceSyncPerpetualTaskInfoDTO instanceSyncPerpetualTaskInfoDTO, boolean isInstanceSyncV2) {
     instanceSyncPerpetualTaskInfoService.deleteById(
         instanceSyncPerpetualTaskInfoDTO.getAccountIdentifier(), instanceSyncPerpetualTaskInfoDTO.getId());
+    if (isInstanceSyncV2
+        && isEmpty(instanceSyncPerpetualTaskInfoService.findAll(instanceSyncPerpetualTaskInfoDTO.getAccountIdentifier(),
+            instanceSyncPerpetualTaskInfoDTO.getPerpetualTaskIdV2()))) {
+      instanceSyncPerpetualTaskService.deletePerpetualTask(instanceSyncPerpetualTaskInfoDTO.getAccountIdentifier(),
+          instanceSyncPerpetualTaskInfoDTO.getPerpetualTaskIdV2());
+    }
+    instanceSyncPerpetualTaskService.deletePerpetualTask(
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/helper/InstanceSyncHelper.java:56]: perpetualTaskId can be null

File: 126-instance/src/main/java/io/harness/helper/K8sCloudConfigMetadata.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright 2023 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
+ */
+
+package io.harness.helper;
+
+public interface K8sCloudConfigMetadata {
+  String clusterName();
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/helper/K8sCloudConfigMetadata.java:11]: This should be getClusterName()

File: 126-instance/src/main/java/io/harness/helper/K8sGcpCloudConfigMetadata.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+@OwnedBy(HarnessTeam.CDP)
+public class K8sGcpCloudConfigMetadata implements K8sCloudConfigMetadata {
+  String clusterName;
+  @Override
+  public String clusterName() {
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/helper/K8sGcpCloudConfigMetadata.java:22]: Once you rename the method to getClusterName() you wouldn't need this override

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java [hunk 9]
```diff
@@ -91,12 +97,29 @@ public DeploymentReleaseDetails getDeploymentReleaseDetails(
...
         K8sDeploymentInfoDTO k8sDeploymentInfoDTO = (K8sDeploymentInfoDTO) deploymentInfoDTO;
-        k8sDeploymentReleaseDetailsList.add(K8sDeploymentReleaseDetails.builder()
-                                                .releaseName(k8sDeploymentInfoDTO.getReleaseName())
-                                                .namespaces(k8sDeploymentInfoDTO.getNamespaces())
-                                                .build());
+        String subscriptionId = null;
+        String resourceGroup = null;
+        boolean useClusterAdminCredentials = false;
+        String clusterName = k8sDeploymentInfoDTO.getCloudConfigMetadata().clusterName();
+        if (k8sDeploymentInfoDTO.getCloudConfigMetadata() instanceof K8sAzureCloudConfigMetadata) {
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesynchandler/K8sInstanceSyncHandler.java:105]: This logic and the from infrastructure outcome is better to move to a utility class

File: 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                                   .clusterName(clusterName)
+                                   .subscriptionId(subscriptionId)
+                                   .resourceGroup(resourceGroup)
+                                   .useClusterAdminCredentials(useClusterAdminCredentials)
+                                   .build())
+        .build();
+  }
+
+  public K8sCloudConfigMetadata getK8sCloudConfigMetadata(InfrastructureOutcome infrastructureOutcome) {
+    if (infrastructureOutcome instanceof K8sGcpInfrastructureOutcome) {
```
Comment (abhinavsingh1196) [inline 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java:55]: a good way to get rid of a lot of these `if (a instanceof b)` checks is to write a separate interface `K8sInfraDataEmitter` which has this method:
`K8sCloudConfigMetadata getInfraOutcomeMetadata`

All the necessary outcomes(like `K8sAwsInfrastructureOutcome`) can implement this interface and return the necessary data. We just have to call that method here then.

File: 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+import io.harness.cdng.infra.beans.K8sGcpInfrastructureOutcome;
+import io.harness.dtos.deploymentinfo.DeploymentInfoDTO;
+import io.harness.dtos.deploymentinfo.K8sDeploymentInfoDTO;
+import io.harness.perpetualtask.instancesync.k8s.K8sDeploymentReleaseDetails;
+import io.harness.perpetualtask.instancesync.k8s.KubernetesCloudClusterConfig;
+
+import lombok.extern.slf4j.Slf4j;
+
+@OwnedBy(HarnessTeam.CDP)
+@Slf4j
```
Comment (abhinavsingh1196) [inline 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java:24]: should add `@Singleton` annotation here

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java (hunk 11 above)
Comment (abosii-harness) [reply 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java:24]: I would say it should be @UtilityClass

**Comment by namantalaycha10** [comment]
trigger uts

File: 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java (hunk 11 above)
Comment (abhinavsingh1196) [reply 126-instance/src/main/java/io/harness/helper/K8sInfrastructureUtility.java:22]: yes my bad, I see no state is being altered by it. Should be a Utility class.

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]

**Review by abosii-harness — APPROVED** [review]

**Comment by namantalaycha10** [comment]
trigger review

**Review by TarunUba1 — APPROVED** [review]

**Pull request merged**